### PR TITLE
[webui] fix repository cycle view

### PR DESCRIPTION
**Pull request by adrianschroeter**
* old code did not handle the fact when a package was in multiple
  cycles and did not show these cycles at all.
* new code still tries to find groups, but ensures that all packages got listed

some code is enabled in test mode only. Not so useful there because it
only makes sense with larger cycle information, but I want to keep
the code for debugging purposes.

### CODE REVIEW

**Comment by adrianschroeter** [comment]
relative high prio since distro people need to deal with new backend cycle handling code.

**Comment by adrianschroeter** [comment]
mls had a better idea over night, will adapt.

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 1]
```diff
@@ -235,37 +235,46 @@ def calculate_repo_cycle(arch)
...
-      list = []
-      packages.each do |package, cycle|
-        list.push(package) if cycle == i
+
+      # take only the first item of each cycle, the rest are duplicates since they are identical
+      packagecycles.keys.map { |p| packagecycles[p][0] }.sort.uniq.each do |key|
+        cycles << packagecycles[key]
+      end
+
+      # self check code, verify that we did not lost any item
```
Comment (mschnitzer) [inline src/api/app/controllers/webui/repositories_controller.rb]: [nitpick] "did not lost" -> "did not lose"

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 2]
```diff
@@ -235,37 +235,46 @@ def calculate_repo_cycle(arch)
...
+      if Rails.env.test?
```
Comment (mschnitzer) [inline src/api/app/controllers/webui/repositories_controller.rb]: Why only in the test environment?

**Comment by adrianschroeter** [comment]
On Donnerstag, 18. Januar 2018, 07:29:07 CET wrote Manuel Schnitzer:
> mschnitzer commented on this pull request.
> 
> 
> 
> >          end
>        end
> -    end
> -    cycles = []
> -    1.upto(nr_cycles) do |i|
> -      list = []
> -      packages.each do |package, cycle|
> -        list.push(package) if cycle == i
> +
> +      # take only the first item of each cycle, the rest are duplicates since they are identical
> +      packagecycles.keys.map { |p| packagecycles[p][0] }.sort.uniq.each do |key|
> +        cycles << packagecycles[key]
> +      end
> +
> +      # self check code, verify that we did not lost any item
> +      if Rails.env.test?
> 
> Why only in the test environment?

because it raises an exception though it leads to a UI crash. 
And it takes time, so it makes not so much sense to run always.

I don't want to loose the code for debugging purposes though.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

File: src/api/app/controllers/webui/repositories_controller.rb [hunk 3]
```diff
@@ -235,37 +235,46 @@ def calculate_repo_cycle(arch)
...
-      packages.each do |package, cycle|
-        list.push(package) if cycle == i
+
+      # take only the first item of each cycle, the rest are duplicates since they are identical
+      packagecycles.keys.map { |p| packagecycles[p][0] }.sort.uniq.each do |key|
+        cycles << packagecycles[key]
+      end
+
+      # self check code, verify that we did not lose any item
+      if Rails.env.test?
```
Comment (hennevogel) [inline src/api/app/controllers/webui/repositories_controller.rb]: So you want to enable/disable this when debugging? Why not put this behind a feature flag instead? Something that you can enable/disable during runtime...

**Comment by hennevogel** [comment]
Do you know of a project with cycles or do you have an example of `BuilddepInfo` output with a cycle?

**Comment by adrianschroeter** [comment]
On Donnerstag, 18. Januar 2018, 14:41:38 CET wrote Henne Vogelsang:
> Do you know of a project with cycles or do you have an example of `BuilddepInfo` output with a cycle?

All base projects, eg. openSUSE:Factory

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by adrianschroeter** [comment]
On Donnerstag, 18. Januar 2018, 14:34:22 CET wrote Henne Vogelsang:
> hennevogel requested changes on this pull request.
> 
> 
> 
> >          end
>        end
> -    end
> -    cycles = []
> -    1.upto(nr_cycles) do |i|
> -      list = []
> -      packages.each do |package, cycle|
> -        list.push(package) if cycle == i
> +
> +      # take only the first item of each cycle, the rest are duplicates since they are identical
> +      packagecycles.keys.map { |p| packagecycles[p][0] }.sort.uniq.each do |key|
> +        cycles << packagecycles[key]
> +      end
> +
> +      # self check code, verify that we did not lose any item
> +      if Rails.env.test?
> 
> So you want to enable/disable this when debugging? Why not put this behind a feature flag instead? Something that you can enable/disable during runtime...

You never want to enable that in production since it creates an exception IMHO.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> You never want to enable that in production since it creates an exception IMHO.

So what's the condition that you want to use this code path then?

**Comment by adrianschroeter** [comment]
On Donnerstag, 18. Januar 2018, 17:13:19 CET wrote Henne Vogelsang:
> > You never want to enable that in production since it creates an exception IMHO.
> 
> So what's the condition that you want to use this code path then?

I just want to avoid to loose it for being able to debug reports fast in first place.

We can drop here, but where should we put it then?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> We can drop here, but where should we put it then?

I don't know because I don't understand what condition has to be fulfilled that you want to use this code path...

**Comment by adrianschroeter** [comment]
On Mittwoch, 31. Januar 2018, 17:42:21 CET wrote Henne Vogelsang:
> > We can drop here, but where should we put it then?
> 
> I don't know because I don't understand what condition has to be fulfilled that you want to use this code path...

For debugging as said before.

Some reports scheduling issues in real-life project X and you need to
check why the webui lacks critial information.

How would you do that without this code?
(and as usual these kinds of things need to be handled under time pressure,
 because product release X depends on it in next hour)

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> Some reports scheduling issues in real-life project X and you need to check why the webui lacks critial information.  How would you do that without this code?

Put it behind a feature switch that you enable if you are in this situation. So you don't change deployed code and the code doesn't behave differently in the test suite.

**Comment by adrianschroeter** [comment]
again, a feature switch is what you never want due to the exception.

But I am just removing the code now if you never want to debug the algorith.

**Comment by hennevogel** [comment]
> But I am just removing the code now if you never want to debug the algorithm.

Okay cool.

@mdeniz and me tried to understand what you are doing here but failed, given the output of `BuilddepInfo` for Factory. Would be cool if you or @mlschroe could enlighten us.

**Comment by adrianschroeter** [comment]
On Donnerstag, 1. Februar 2018, 18:32:43 CET wrote Henne Vogelsang:
> > But I am just removing the code now if you never want to debug the algorithm.
> 
> Okay cool.
> 
> @mdeniz and me tried to understand what you are doing here but failed, given the output of `BuilddepInfo` for Factory. Would be cool if you or @mlschroe could enlighten us. 

that would be indeed the more important discussion instead of the cosmetic part;)

In short the algorithm brings all cycles together which have an overlap of (at least)
one package. Since bootstrapping this cycle will influence also the other cycle, so
you have one bigger cycle.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Pull request merged**